dj/api/helpers.py

```diff
@@ -234,30 +246,85 @@ def validate_node_data(
+                        debug={"missing_parents": list(missing_parents_map.keys())},
+                    ),
+                ],
+            )
 
     # Add aliases for any unnamed columns and confirm that all column types can be inferred
     query_ast.select.add_aliases_to_unnamed_columns()
-    node_revision.columns = [
-        Column(name=col.name.name, type=get_type_of_expression(col))  # type: ignore
-        for col in query_ast.select.projection
-    ]
 
-    node_revision.status = NodeStatus.VALID
-    return node, node_revision, dependencies_map
+    validated_node.columns = []
+    type_inference_failed_columns = []
+    for col in query_ast.select.projection:
+        try:
+            column_type = get_type_of_expression(col)
+            validated_node.columns.append(
+                Column(name=col.name.name, type=get_type_of_expression(col)),
+            )
+        except DJParseException:  # p
+            type_inference_failed_columns.append(col.name.name)
+            validated_node.status = NodeStatus.INVALID
+
+    return (
+        validated_node,
+        dependencies_map,
+        missing_parents_map,
+        type_inference_failed_columns,
+    )
+
+
+def resolve_downstream_references(session: Session, node: NodeRevision) -> int:
+    """
+    Find all nodes with missing parent references to `node` and resolve them
+    """
+    downstream_references_updated = 0
+    missing_parents = session.exec(
+        select(MissingParent).where(MissingParent.name == node.name),
+    ).all()
+    for missing_parent in missing_parents:
+        missing_parent_links = session.exec(
+            select(NodeMissingParents).where(
+                NodeMissingParents.missing_parent_id == missing_parent.id,
+            ),
+        ).all()
+        downstream_node_ids = [
+            node.referencing_node_id for node in missing_parent_links
+        ]
+        for id in downstream_node_ids:  # Remove from missing parents and add to parents
+            downstream_node_revision = (
+                session.exec(select(NodeRevision).where(NodeRevision.id == id))
+                .unique()
+                .one()
+            )
+            downstream_node_revision.parents.append(node)
+            downstream_node_revision.missing_parents.remove(missing_parent)
+            (
+                _,
+                _,
+                missing_parents_map,
+                type_inference_failed_columns,
+            ) = validate_node_data(data=downstream_node_revision, session=session)
+            if not missing_parents_map and not type_inference_failed_columns:
+                downstream_node_revision.status = NodeStatus.VALID
+            session.add(downstream_node_revision)
+            downstream_references_updated += 1
+
+        session.delete(missing_parent)  # Remove missing parent reference to node
+    return downstream_references_updated
```

Comment:
Is this value just for auditing? As in, we don't actually use this count anywhere right?

Reply:
Yeah good catch, I really just returned that because otherwise the function felt too opaque to the calling logic. I was thinking we could include it with the metadata that's returned when you create a node, that way you get at least some information it's had an effect to the downstream portion of the DAG. In the future we can enrich that even more by specifying which nodes were specifically updated and how.

Reply:
Changed this to return a list of nodes that have been newly switched from invalid to valid. That's then used to know how to propagate the effect and revalidate all downstream nodes.